ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/API.hs

```diff
@@ -565,12 +562,47 @@ data AnnLedgerError m l blk = AnnLedgerError {
     , annLedgerErr    :: LedgerErr l
     }
 
+type AnnLedgerError' m blk = AnnLedgerError m (ExtLedgerState blk) blk
+
+{-------------------------------------------------------------------------------
+  Finding blocks
+-------------------------------------------------------------------------------}
+
+-- | Resolve a block
+--
+-- Resolving a block reference to the actual block lives in @m@ because
+-- it might need to read the block from disk (and can therefore not be
+-- done inside an STM transaction).
+--
+-- NOTE: The ledger DB will only ask the 'ChainDB' for blocks it knows
+-- must exist. If the 'ChainDB' is unable to fulfill the request, data
+-- corruption must have happened and the 'ChainDB' should trigger
+-- validation mode.
+type ResolveBlock m blk = RealPoint blk -> m blk
+
+-- | Monads in which we can resolve blocks
+--
+-- To guide type inference, we insist that we must be able to infer the type
+-- of the block we are resolving from the type of the monad.
+class Monad m => ResolvesBlocks m blk | m -> blk where
+  doResolveBlock :: ResolveBlock m blk
+
+instance Monad m => ResolvesBlocks (ReaderT (ResolveBlock m blk) m) blk where
+  doResolveBlock r = ReaderT $ \f -> f r
+
+-- Quite a specific instance so we can satisfy the fundep
+instance Monad m
+      => ResolvesBlocks (ExceptT e (ReaderT (ResolveBlock m blk) m)) blk where
+  doResolveBlock = lift . doResolveBlock
+
```

Comment:
Technically not part of the API, but part of initialisation (IMO that's impl-specific)